File: tmw-aspectj/src/main/java/com/tmw/gradle/aspectj/utils/FilterUtil.java
package com.tmw.gradle.aspectj.utils;

import java.util.ArrayList;
import java.util.List;

/**
 * Aspect默认过滤
 *
 * @author tanmingwu848
 * @since 2021/09/15
 */
public class FilterUtil {

    private static final List<String> CLASS_FILTERS = new ArrayList<>();
    private static final List<String> METHOD_FILTERS = new ArrayList<>();
    private static final String SPECIAL = "$";

    static {
        CLASS_FILTERS.add("BuildConfig.class");
        CLASS_FILTERS.add("R.class");
        CLASS_FILTERS.add("R$");

        METHOD_FILTERS.add("<init>");
        METHOD_FILTERS.add("<clinit>");
        METHOD_FILTERS.add("invoke");
    }

    private static boolean isFilter(String name) {
        for (String tag : CLASS_FILTERS) {
            if (name.contains(tag)) {
                return true;
            }
        }
        return false;
    }

    public static boolean isAvailable(String className) {
        if (isFilter(className)) {
            return false;
        }

        return className.endsWith(".class");
    }

    public static boolean isMethodAvailable(String methodName) {
        return !METHOD_FILTERS.contains(methodName) /*&& !methodName.contains(SPECIAL)*/;
    }


}

File: tmw-aspectj/src/main/java/com/tmw/gradle/aspectj/utils/FilterUtil.kt
package com.tmw.gradle.aspectj.utils

import com.android.build.api.transform.JarInput
import java.io.IOException
import java.util.jar.JarEntry
import java.util.jar.JarFile

/**
 * Aspect默认过滤
 *
 * @author tanmingwu
 * @since 2021/09/15
 */
object FilterUtil {

    private val CLASS_FILTERS: MutableList<String> = arrayListOf(
        "R.jar"
    )

    private fun isFilter(name: String): Boolean {
        return CLASS_FILTERS.contains(name)
    }

    fun isExclude(jarInput: JarInput, excludes: List<String>): Boolean {
        try {
            val firstClass = getFirstClass(jarInput)
            firstClass ?: return false
            if (isFilter(firstClass.name)) {
                return true
            }
            return ListUtil.contain(firstClass.name, excludes)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    fun isInclude(jarInput: JarInput, includes: List<String>): Boolean {
        try {
            if (isFilter(jarInput.file.name)) {
                return false
            }

            val firstClass = getFirstClass(jarInput)
            firstClass ?: return false
            return ListUtil.contain(firstClass.name, includes)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    private fun getFirstClass(jarInput: JarInput): JarEntry? {
        if (!jarInput.file.absolutePath.endsWith(".jar")) {
            return null
        }
        val jarFile = JarFile(jarInput.file)
        val enumeration = jarFile.entries()
        if (!enumeration.hasMoreElements()) {
            return null
        }
        var first = enumeration.nextElement()
        while (!first.name.endsWith(".class") && enumeration.hasMoreElements()) {
            first = enumeration.nextElement()
        }
        return first
    }

    fun checkAspectPath(jarInput: JarInput): Boolean {
        val firstClass = getFirstClass(jarInput)
        firstClass ?: return false
        if (firstClass.name.contains("com/example/baselibrary")) {
            return true
        }
        return false
    }
}